lib/reversetunnel/resolver.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reversetunnel
+
+import (
+	"context"
+
+	"github.com/gravitational/teleport/api/client/webclient"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+)
+
+// Resolver looks up reverse tunnel addresses
+type Resolver func() (*utils.NetAddr, error)
```

Comment:
I'm a bit concerned about the increased network load if proxies are restarted in very large clusters.  Might create another thundering herd problem if we reload the address on every call to the resolver.  I'm thinking it might be good if the resolver had some basic ttl-based caching capabilities, so that if it were called many times over a small period (say 3-5 seconds), only one network call is actually made.

We already use [lib/cache/fncache.go](https://github.com/gravitational/teleport/blob/b384de600791133e72d0849432528e9d3969218b/lib/cache/fncache.go) to do the same thing to reduce thundering herd issues when the cache is unhealthy.  That helper doesn't rely on anything in `lib/cache`, so we could easily move it somewhere under `utils`, and use it to provide similar functionality here.  Ex:

```golang
func CachingResolver(resolver Resolver) Resolver {
    cache := utils.NewFnCache(3 * time.Second)
    return func() (*utils.NetAddr, error) {
        a, err := cache.Get(context.TODO(), "resolver", func() (interface{}, error) {
            addr, err := resolver()
            return addr, err
        })
        if err != nil {
            return nil, err
        }
        return a.(*utils.NetAddr), nil
    }
}
```